### PR TITLE
Add initial version of the Cloud workbench

### PR DESCRIPTION
**Pull request by vejmarie**
This is adding initial Object storage support to S3 type of Cloud storage. As to use it just use the basic feature

import Cloud
Cloud.cloudurl(u"http://127.0.0.1") -> Your s3 cloud server IP
Cloud.cloudtcpport(u"9000") -> TCP port of the server
Cloud.cloudaccesskey(u"588N07RGWPGZXHEM1YWA") -> Access Key
Cloud.cloudsecretkey(u"vBdvcG8uIk3s9A1DtL82beWI+V0LYW+cPDES+CLn") -> Secret Key
Cloud.cloudrestore(u"test") -> Name of the file to restore (bucket name)
Cloud.cloudsave(u"test") -> Save a file into a bucket

### CODE REVIEW

**Comment by yorikvanhavre** [comment]
That looks really cool and elegantly implemented! I prefer @wwmayer to have a look, though, as it touches the core files. Just a detail, could you fix the license headers (some missing and some anonymous)? Otherwise we always get annoying complaints at debian...

**Comment by vejmarie** [comment]
I will update the license in a couple of hours. Does the one in the AppCloud.cpp is ok ? (I do plan just to copy and paste it ;)). Thanks for your feedback, and let's cross fingers to get that initial PR going through. I do have plan to implement some XMPP like stuff on top of it, to ease collaborative work

File: src/Mod/Cloud/Gui/Command.cpp [hunk 1]
```diff
@@ -0,0 +1,63 @@
+/***************************************************************************
+ *   Copyright (c) YEAR YOUR NAME         <Your e-mail address>            *
```
Comment (luzpaz) [inline src/Mod/Cloud/Gui/Command.cpp]: Missing info here

File: src/Mod/Cloud/Gui/PreCompiled.cpp [hunk 2]
```diff
@@ -0,0 +1,24 @@
+/***************************************************************************
+ *   Copyright (c) YEAR YOUR NAME         <Your e-mail address>            *
```
Comment (luzpaz) [inline src/Mod/Cloud/Gui/PreCompiled.cpp]: Missing info here

File: src/Mod/Cloud/Gui/PreCompiled.h [hunk 3]
```diff
@@ -0,0 +1,70 @@
+/***************************************************************************
+ *   Copyright (c) YEAR YOUR NAME         <Your e-mail address>            *
```
Comment (luzpaz) [inline src/Mod/Cloud/Gui/PreCompiled.h]: Missing info here

File: src/Mod/Cloud/Gui/Workbench.cpp [hunk 4]
```diff
@@ -0,0 +1,65 @@
+/***************************************************************************
+ *   Copyright (c) YEAR YOUR NAME         <Your e-mail address>            *
```
Comment (luzpaz) [inline src/Mod/Cloud/Gui/Workbench.cpp]: Missing info here

File: src/Mod/Cloud/Gui/Workbench.h [hunk 5]
```diff
@@ -0,0 +1,47 @@
+/***************************************************************************
+ *   Copyright (c) YEAR YOUR NAME         <Your e-mail address>            *
```
Comment (luzpaz) [inline src/Mod/Cloud/Gui/Workbench.h]: Missing info here

**Comment by luzpaz** [comment]
Heads-up @kkremitzki, any issue with downstream debian with this new code ?

**Comment by kkremitzki** [comment]
> Cloud.cloudurl(u"http://127.0.0.1") -> Your s3 cloud server IP

Still on Python 2? :laughing:

File: src/App/Document.cpp [hunk 6]
```diff
@@ -1261,6 +1261,52 @@ bool Document::isTransactionEmpty() const
...
+    this->d->clearRecomputeLog();
+    this->d->objectArray.clear();
+    this->d->objectMap.clear();
+    this->d->objectIdMap.clear();
+    this->d->lastObjectId = 0;
+}
+
+void Document::resetTouched()
+{
+/*
```
Comment (kkremitzki) [inline src/App/Document.cpp]: Can we do something about this commented out section?

File: src/Mod/Cloud/App/AppCloud.cpp [hunk 7]
```diff
@@ -0,0 +1,693 @@
...
+                struct curl_slist *chunk = NULL;
+                char header_data[1024];
+                // Let's build our own header
+                sprintf(header_data,"Host: %s:%s", strtok((char *)this->Url,"http://"),
+                                                   this->TcpPort);
+                chunk = curl_slist_append(chunk, header_data);
+                sprintf(header_data,"Date: %s", date_formatted);
+                chunk = curl_slist_append(chunk, header_data);
+                chunk = curl_slist_append(chunk, "Content-Type: application/xml");
+               std::string digest_str;
```
Comment (kkremitzki) [inline src/Mod/Cloud/App/AppCloud.cpp]: Missing a space on the indentation here.

File: src/Mod/Cloud/Gui/PreCompiled.cpp (hunk 2 above)
Comment (vejmarie) [reply src/Mod/Cloud/Gui/PreCompiled.cpp]: Fixed thanks

File: src/Mod/Cloud/Gui/PreCompiled.h (hunk 3 above)
Comment (vejmarie) [reply src/Mod/Cloud/Gui/PreCompiled.h]: Shall be fixed now

File: src/Mod/Cloud/Gui/Workbench.cpp (hunk 4 above)
Comment (vejmarie) [reply src/Mod/Cloud/Gui/Workbench.cpp]: Shall be fixed now

File: src/App/Document.cpp (hunk 6 above)
Comment (vejmarie) [reply src/App/Document.cpp]: You are right, that is dead code coming from the fact that between the time I wrote the code and pushed for the PR, there has been updates into document management. I removed that code and calls

File: src/Mod/Cloud/App/AppCloud.cpp (hunk 7 above)
Comment (vejmarie) [reply src/Mod/Cloud/App/AppCloud.cpp]: I have updated the file, thanks to have pointed it out

**Comment by vejmarie** [comment]
> > Cloud.cloudurl(u"http://127.0.0.1") -> Your s3 cloud server IP
> 
> Still on Python 2? 😆

Yeah ! But using Qt5. I am on Mac, Python 3 and 2 cohabitation is a little bit bumpy, so I will focus on the switch after we have integrated that PR :)

**Comment by wwmayer** [comment]
Two things:
Inside Document::RestoreDocFile the XMLReader is created on the heap and passed to the Base::Reader. However at no point the object is destroyed so that we have a memory leak there.
But in the future some use cases may appear where it's not clear who is responsible to manage the XMLReader instance so that instead of using raw pointers we should use shared_ptr.
Using a shared_ptr will also increase exception-safety.

emptyDocument() should be renamed to clearDocument() (my personal preference) because this makes it more consistent with method names of STL containers. There empty() is used to check if a container has any elements, not to clear it.

I will make these changes when merging the PR but before I have to read through the discussion on the forum again where realthunder proposed a solution. So, this may still take a couple of days...

**Comment by vejmarie** [comment]
> Two things:
> Inside Document::RestoreDocFile the XMLReader is created on the heap and passed to the Base::Reader. However at no point the object is destroyed so that we have a memory leak there.
> But in the future some use cases may appear where it's not clear who is responsible to manage the XMLReader instance so that instead of using raw pointers we should use shared_ptr.
> Using a shared_ptr will also increase exception-safety.
> 

I just made the change, so if the  localreader ptr is not null when the Reader object is destroyed, its content is deleted.

> I will make these changes when merging the PR but before I have to read through the discussion on the forum again where realthunder proposed a solution. So, this may still take a couple of days...

Ok. I didn't write the emptyDocument code, but agree that a clearDocument naming seems more understandable. 

Realthunder proposale is good, but as he is still working on his branch, I focused on trying to make very slight modification to the Document to make the Cloud workbench initial commit getting upstream. I still have a lot of work to do (like error handling with the curl code). 

I could have re-integrated part of his work, but this is his work ;), and he deserves the recognition, and the modifications are wide, which means that it is probably better to keep it as a full PR. I will re-adapt my code after he pushed his PR without too many issues (even if I bet I will have to keep the approach as the Reader I am build is currently external to the core classes)

File: src/Mod/Cloud/App/AppCloud.cpp [hunk 8]
```diff
@@ -0,0 +1,692 @@
...
+                                  (const char*)this->TcpPort.getStrValue().c_str(),
+                                  BucketName);
+
+        mywriter.putNextEntry("Document.xml");
+
+        if (hGrp->GetBool("SaveBinaryBrep", false))
+            mywriter.setMode("BinaryBrep");
+        mywriter.Stream() << "<?xml version='1.0' encoding='utf-8'?>" << endl
+                        << "<!--" << endl
+                        << " FreeCAD Document, see http://www.freecadweb.org for more information..." << endl
```
Comment (luzpaz) [inline src/Mod/Cloud/App/AppCloud.cpp]: please use `https://` here

File: src/Mod/Cloud/Gui/Command.cpp [hunk 9]
```diff
@@ -0,0 +1,63 @@
+/***************************************************************************
+ *   Copyright (c) 2019         jmverdun3@gmail.com	                   *
```
Comment (luzpaz) [inline src/Mod/Cloud/Gui/Command.cpp]: please add you name before the email and fix whitespace after it

File: src/Mod/Cloud/Gui/PreCompiled.cpp [hunk 10]
```diff
@@ -0,0 +1,24 @@
+/***************************************************************************
+ *   Copyright (c) 2019         jmverdun3@gmail.com	                   *
```
Comment (luzpaz) [inline src/Mod/Cloud/Gui/PreCompiled.cpp]: please add you name before the email and fix whitespace after it

File: src/Mod/Cloud/Gui/PreCompiled.h [hunk 11]
```diff
@@ -0,0 +1,70 @@
+/***************************************************************************
+ *   Copyright (c) 2019         jmverdun3@gmail.com                        *
```
Comment (luzpaz) [inline src/Mod/Cloud/Gui/PreCompiled.h]: please add you name before the email and fix whitespace after it

File: src/Mod/Cloud/Gui/Workbench.cpp [hunk 12]
```diff
@@ -0,0 +1,65 @@
+/***************************************************************************
+ *   Copyright (c) 2019         jmverdun3@gmail.com                        *
```
Comment (luzpaz) [inline src/Mod/Cloud/Gui/Workbench.cpp]: please add you name before the email and fix whitespace after it

File: src/Mod/Cloud/Gui/Workbench.h [hunk 13]
```diff
@@ -0,0 +1,47 @@
+/***************************************************************************
+ *   Copyright (c) 2019         jmverdun3@gmail.com                        *
```
Comment (luzpaz) [inline src/Mod/Cloud/Gui/Workbench.h]: please add you name before the email and fix whitespace after it

File: src/Mod/Cloud/App/AppCloud.cpp (hunk 8 above)
Comment (vejmarie) [reply src/Mod/Cloud/App/AppCloud.cpp]: Shall be fixed

File: src/Mod/Cloud/Gui/PreCompiled.cpp (hunk 10 above)
Comment (vejmarie) [reply src/Mod/Cloud/Gui/PreCompiled.cpp]: Shall be fixed

File: src/Mod/Cloud/Gui/PreCompiled.h (hunk 11 above)
Comment (vejmarie) [reply src/Mod/Cloud/Gui/PreCompiled.h]: shall be fixed

File: src/Mod/Cloud/Gui/Workbench.h (hunk 13 above)
Comment (vejmarie) [reply src/Mod/Cloud/Gui/Workbench.h]: shall be fixed

File: src/Mod/Cloud/Gui/Command.cpp (hunk 9 above)
Comment (vejmarie) [reply src/Mod/Cloud/Gui/Command.cpp]: shall be fixed

File: src/Mod/Cloud/Gui/Workbench.cpp (hunk 12 above)
Comment (vejmarie) [reply src/Mod/Cloud/Gui/Workbench.cpp]: shall be fixed

**Comment by vejmarie** [comment]
Is there something broken with the windows build or did I introduced something bad ?

**Comment by luzpaz** [comment]
@vejmarie JFYI, can we also make a note to add some doxygen comments so that this code shows up in the API files ? User @vocx-fc re-wrote the [Doxygen docs on the 
 FC wiki](https://www.freecadweb.org/wiki/Doxygen) to help facilitate better comprehension of the code by other devs. Thanks for your consideration in advance.

**Comment by wwmayer** [comment]
Merged.

Fixed Win32 build failures in d223e326ce0 but to get a string from time info still must be done.
I wonder whether using std of C++, boost or Qt instead of plain C would be the better option.

**Pull request closed without merging**